tree: introduce concept of "default" collation

otan opened the pull request:
Resolves https://github.com/cockroachdb/cockroach/issues/54989


Release note (sql change): Introduced a pg_collation of "default".
Strings now return the "default" collation OID in the pg_attribute
table (this was previously en_US). The "default" collation is also
visible on the pg_collation virtual table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56598)
<!-- Reviewable:end -->

otan committed:
tree: introduce concept of "default" collation

Release note (sql change): Introduced a pg_collation of "default".
Strings now return the "default" collation OID in the pg_attribute
table (this was previously en_US). The "default" collation is also
visible on the pg_collation virtual table.

rafiss commented on pkg/sql/logictest/testdata/logic_test/pg_catalog:
```diff
@@ -1358,14 +1358,14 @@ ORDER BY oid
 17      bytea          0         0             NULL           NULL        NULL
-18      char           0         3903121477    NULL           NULL        NULL
-19      name           0         3903121477    NULL           NULL        NULL
+18      char           0         3403232968    NULL           NULL        NULL
```
seems like `char` in PG has a typcollation of 0

otan replied:
hmm, i think this is a separate issue in that this is what pg_type reports, which is different.
i think that's a separate issue. we currently base this column on the collation of the type...

rafiss commented on pkg/sql/logictest/testdata/logic_test/pg_catalog:
```diff
@@ -1358,14 +1358,14 @@ ORDER BY oid
-18      char           0         3903121477    NULL           NULL        NULL
-19      name           0         3903121477    NULL           NULL        NULL
+18      char           0         3403232968    NULL           NULL        NULL
+19      name           0         3403232968    NULL           NULL        NULL
```
`name` has a collation of `C` in PG

???

otan replied:
yeah i think that makes sense as name is an internal thing isn't it? either way we don't have C, so i think we can leave this as is

rafiss commented on pkg/sql/sem/tree/create.go:
```diff
@@ -460,15 +460,21 @@ func NewColumnTableDef(
+			// To most behave like postgres, set the CollatedString type if a non-"default"
+			// collation is used.
+			if locale != DefaultCollationTag {
+				_, err := language.Parse(locale)
```
so i guess here would be the place where we could normalize the collation name using `lex.EncodeLocaleName`. do you think we should do that?

(either in this PR, or at any point?)

otan replied:
yeah, as well as change all `language.Parse` calls to ensure it does this appropriately...

otan replied:
yeah, that sounds about right. maybe somewhere with alter column type as well.
maybe in general we shouldn't use language.Parse but write our own wrapper around it.

rafiss reviewed (commented):
lgtm -- but wondering about `char` and `name`

otan commented:
char and name seems like an existing problem based on the type descriptor -- so i'm going to ignore this for now.

https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/pg_catalog.go#L2276

going to leave this to another issue.

TFTR!

bors r=rafiss
